app/modules/tmr.c

```diff
@@ -132,26 +132,29 @@ static int tmr_start(lua_State* L){
 	luaL_argcheck(L, lua_isboolean(L, 2) || lua_isnil(L, 2), 2, "boolean expected");
 	int restart = lua_toboolean(L, 2);
 
-	lua_settop(L, 1);  /* ignore any args after the userdata */
+	lua_settop(L, 1);  /* we need to have userdata on top of the stack */
```

Comment:
This is only needed because tmr_start() is a `LUA_API` function but is also called directly by another `LUA_API` function tmr_alarm(), which is a bit naughty.  Since the alarm function is doing the naughties, then it should get the stack right, and this settop should be moved to the line immediately above the  `trm_start(L)` call in `trm_alarm()`. 

According to the documentation:
-  `start` is supposed to return `nil` if the timer is not registered and  (`true`, `mode`) otherwise. **yet in reality it only returns the boolean (was the timer started?) 
-  `alarm' is supposed `true` if the timer was started and `false` on error.

So `t:start(true)` is supposed to force a restart, but what is the behaviour of `t:alarm()` if the timer is already running?  Do you need the restart flag here as well.

Either way, the documentation should reflect the actual behaviour.

Reply:
@TerryE I already have `lua_settop` in `tmr_alarm()` at line [155-157](https://github.com/nodemcu/nodemcu-firmware/pull/3263/commits/cbf2afa7caf657e0ddcc6fddc4a2cbc070157d20#diff-ac89d4cec75b9b260a2fe9e732d2c50bR155-R157).

This one `lua_settop` is there because we need the `tmr.timer` UD on the top of the stack before eventually calling 
```
tmr->self_ref = luaL_ref(L, LUA_REGISTRYINDEX);
``` 
This gets rid off the optional argument.

Do you agree with this reasoning and approach?

Reply:
Can we decouple `tmr_start()` and `tmr_alarm()` so that their `LUA_API` aspects are separate (but they might use common bits of C code encapsulated as functions)?

Reply:
Do you find this still too dirty?

I think it would not be very elegant as in `tmr_register` only one line is C code (`os_timer_setfn(&tmr->os, alarm_timer_common, tmr);`) and 16 lines relates to `LUA_API`.
It's very similar for `tmr_start`. 

To me this implementation is not so much dirty and if documented well in the code it is readable and does the job well.

Reply:
Lukas, the LUA_API entries are really intended to be invoked through `lua_call()`  If we are calling one directly then this should be clearly documented inline. _Either_ the caller or the callee should properly balance the Lua stack inbound and outbound.

Since we document the behaviour of start with the optional flag, the `tobj:alarm()` documentation should explain what happens is a `tobj` is an exisitng timer has been started.

Reply:
> the LUA_API entries are really intended to be invoked through lua_call() If we are calling one directly then this should be clearly documented inline. Either the caller or the callee should properly balance the Lua stack inbound and outbound.

This is clear. I think the proposed implementation does exactly that.

> Since we document the behaviour of start with the optional flag, the tobj:alarm() documentation should explain what happens is a tobj is an existing timer has been started.

Do you think we should add optional parameter `restart` to `tmr.alam()`? That would be probably the way giving the highest degree of freedom to user. It should be easy to be implemented + documented inline as well as in documentation.

Reply:
> I think this is the right behavior.

I went through the documentation and checked the obj methods again for consistency (primarily behavioral). It all makes sense _to me_.